### PR TITLE
Fix Out of bounds read due to lack of string termination in assert

### PR DESCRIPTION
**Pull request by liqunfu**

### CODE REVIEW

**Review by gramalingam — APPROVED** [review]
minor nit: I was wondering if either 2048 or a named constant would be better than sizeof(msg)? As far as I checked, looks like char is guaranteed to be 1 byte but was wondering, just in case.

File: onnx/common/assertions.cc [hunk 1]
```diff
@@ -21,9 +21,12 @@ std::string barf(const char* fmt, ...) {
   va_list args;
 
   va_start(args, fmt);
-  // Although vsnprintf might have vulnerability issue while using format string with overflowed length,
-  // it should be safe here to use fixed length for buffer "msg". No further checking is needed.
-  vsnprintf(msg, 2048, fmt, args);
+
+  // use fixed length for buffer "msg" to avoid buffer overflow
+  vsnprintf(msg, sizeof(msg) - 1, fmt, args);
```
Comment (justinchuby) [inline onnx/common/assertions.cc:26]: Is this relevant? 

> warning: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay]

Comment (liqunfu) [reply onnx/common/assertions.cc:26]: will os.path.basename(path) avoid 'path' being pointed to location outside of cwd? Here is my experiment:
>>> path="/external_data/../../secret"
>>> import os
>>> path2 = os.path.normpath(path)
>>> path2
'\\secret'
>>> path3 = os.path.basename(path)
>>> path3
'secret'

Comment (liqunfu) [reply onnx/common/assertions.cc]: fixed the error with std::array.

File: onnx/common/assertions.cc [hunk 2]
```diff
@@ -24,7 +24,7 @@ std::string barf(const char* fmt, ...) {
   va_start(args, fmt);
 
   // use fixed length for buffer "msg" to avoid buffer overflow
-  vsnprintf(msg.data(), msg.size() - 1, fmt, args);
+  vsnprintf(static_cast<char*>(msg.data()), msg.size() - 1, fmt, args);
```
Comment (justinchuby) [inline onnx/common/assertions.cc:27]: args too?

warning: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay] ```cpp vsnprintf(static_cast<char*>(msg.data()), msg.size() - 1, fmt, args);

Comment (gramalingam) [reply onnx/common/assertions.cc:27]: How (would we fix it for args)? This is used via a macro for printf style untyped inputs.

**Comment by gramalingam** [comment]
How about a test-case?

File: onnx/common/assertions.cc (hunk 2 above)
Comment (justinchuby) [reply onnx/common/assertions.cc:27]: Not sure. Just saw the lint message. Please feel free to ignore.

File: onnx/common/assertions.cc [hunk 3]
```diff
@@ -18,7 +18,8 @@
 namespace ONNX_NAMESPACE {
 
 std::string barf(const char* fmt, ...) {
-  std::array<char, 2048> msg;
+  constexpr size_t BufferSize = 2048;
```
Comment (justinchuby) [inline onnx/common/assertions.cc:21]: nit: snake case for variables `buffer_size`

**Pull request merged**